v2/batcher_test.go

```diff
@@ -15,129 +15,131 @@ import (
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	mgr := &mockLeaseManager{}
+	mgr.On("RaiseEventsTo", mock.Anything)
+	res := gobatcher.NewSharedResource().
+		WithReservedCapacity(2000).
+		WithSharedCapacity(10000, mgr).
+		WithFactor(1000)
+	batcher := gobatcher.NewBatcher().
+		WithRateLimiter(res)
+	err := batcher.Start(ctx)
+	assert.NoError(t, err, "expecting no errors on startup")
+	watcher := gobatcher.NewWatcher(func(batch []gobatcher.Operation) {})
+	good := gobatcher.NewOperation(watcher, 11000, struct{}{}, false)
+	err = batcher.Enqueue(good)
+	assert.NoError(t, err)
+	bad := gobatcher.NewOperation(watcher, 13000, struct{}{}, false)
+	err = batcher.Enqueue(bad)
+	if err != nil {
+		_ = err.Error() // improves code coverage
+	}
+	assert.Equal(t, gobatcher.TooExpensiveError, err, "expect a too-expensive-error error")
+	mgr.AssertNumberOfCalls(t, "RaiseEventsTo", 1)
+}
 
-	t.Run("operations cannot be attempted more than x times", func(t *testing.T) {
-		// NOTE: this test works by recursively enqueuing the same operation over and over again until it fails
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-		batcher := gobatcher.NewBatcher().
-			WithFlushInterval(1 * time.Millisecond)
-		err := batcher.Start(ctx)
-		assert.NoError(t, err, "expecting no errors on startup")
-		var mu sync.Mutex
-		attempts := 0
-		func() {
-			var op gobatcher.Operation
-			enqueue := func() {
-				mu.Lock()
-				defer mu.Unlock()
-				attempts++
-				if eerr := batcher.Enqueue(op); eerr != nil {
-					if eerr != nil {
-						_ = eerr.Error() // improves code coverage
-					}
-					assert.Equal(t, gobatcher.TooManyAttemptsError, eerr, "expect the error to be too-many-attempts")
-					return
-				}
-				if attempts > 8 {
-					assert.FailNow(t, "the max-attempts governor didn't work, we have tried too many times")
+func TestBatcher_Enqueue_OperationsCannotBeAttemptedMoreThanXTimes(t *testing.T) {
+	// NOTE: this test works by recursively enqueuing the same operation over and over again until it fails
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	batcher := gobatcher.NewBatcher().
+		WithFlushInterval(1 * time.Millisecond)
+	err := batcher.Start(ctx)
+	assert.NoError(t, err, "expecting no errors on startup")
+	var mu sync.Mutex
+	attempts := 0
+	func() {
+		var op gobatcher.Operation
+		enqueue := func() {
+			mu.Lock()
+			defer mu.Unlock()
+			attempts++
+			if eerr := batcher.Enqueue(op); eerr != nil {
+				if eerr != nil {
+					_ = eerr.Error() // improves code coverage
 				}
+				assert.Equal(t, gobatcher.TooManyAttemptsError, eerr, "expect the error to be too-many-attempts")
+				return
 			}
-			watcher := gobatcher.NewWatcher(func(batch []gobatcher.Operation) {
-				enqueue()
-			}).WithMaxAttempts(3)
-			op = gobatcher.NewOperation(watcher, 100, struct{}{}, false)
+			if attempts > 8 {
```

Comment:
Yes. Good catch! Thank you @awkwardindustries .

Reply:
That was intentionally written to be 8, but it honestly doesn't matter. You can even remove that entire if clause and it will still be fine. This was a failsafe to fail the test if the WithMaxAttempts() didn't work. Maybe it was even in there from writing the test before the functionality was written.

Regardless, if you want the cleanest path, it can be removed.